clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -904,27 +904,27 @@ public void onFailure(RuntimeException e) {
         for (Map.Entry<TopicPartition, Long> entry: timestampsToSearch.entrySet()) {
             TopicPartition tp  = entry.getKey();
             Long offset = entry.getValue();
-            Optional<MetadataCache.PartitionInfoAndEpoch> currentInfo = metadata.partitionInfoIfCurrent(tp);
-            if (!currentInfo.isPresent()) {
+            Optional<Metadata.LeaderAndEpoch> leaderAndEpochOpt = metadata.currentLeader(tp);
+            if (!leaderAndEpochOpt.isPresent()) {
                 log.debug("Leader for partition {} is unknown for fetching offset {}", tp, offset);
                 metadata.requestUpdate();
                 partitionsToRetry.add(tp);
-            } else if (currentInfo.get().partitionInfo().leader() == null) {
-                log.debug("Leader for partition {} is unavailable for fetching offset {}", tp, offset);
-                metadata.requestUpdate();
-                partitionsToRetry.add(tp);
```

Comment:
We have unified the two cases.

Reply:
Do we want to tweak the unified log to say `unknown or unavailable` or is that not worth it?

Reply:
I think "unknown" is probably clearer in both cases. "Unavailable" suggests we know the leader, but cannot reach it for some reason. In the logic below, for example, we cannot connect to the leader because we are backing off, so we consider it "unavailable."